### PR TITLE
concurrency: LoggedRLock - logs when lock is waited, and supports expiration

### PR DESCRIPTION
**Pull request by koreno**

### CODE REVIEW

File: easypy/concurrency.py [hunk 1]
```diff
@@ -955,6 +958,77 @@ def _get_my_ident():
...
+
+        log_timer = Timer(expiration=None if timeout < 0 else timeout)
+        while not log_timer.expired:
+            timeout = min(log_timer.remain or self._log_interval, self._log_interval)
+            if self._expiration_timer:
+                timeout = min(self._expiration_timer.remain, timeout)
+            if self._lock.acquire(blocking=True, timeout=timeout):
+                # set the timer after acquiring the lock!
+                self._expiration_timer = self._expiration_timer or Timer(expiration=expiration or self._default_expiration)
+                _logger.debug("%s - acquired", self)
```
Comment (noopti) [inline easypy/concurrency.py]: let's log "acquired by {}" ...

File: easypy/concurrency.py [hunk 2]
```diff
@@ -955,6 +958,77 @@ def _get_my_ident():
...
+        try:
+            owner = threading._active[owner].name
+        except KeyError:
+            pass
+        if owner:
+            return "<{}: owned by <{}>x{} for {}>".format(self._name, owner, count, self._expiration_timer.elapsed)
+        else:
+            return "<{}: unowned>".format(self._name)
+
+    def acquire(self, blocking=True, timeout=-1, expiration=None):
```
Comment (tigrawap) [inline easypy/concurrency.py]: Why both blocking and timeout?  
Is not timeout=None == non blocking and timeout explicitly blocking?

File: easypy/concurrency.py (hunk 1 above)
Comment (koreno) [inline easypy/concurrency.py]: Why?...

File: easypy/concurrency.py [hunk 3]
```diff
@@ -955,6 +958,77 @@ def _get_my_ident():
...
+    def __init__(self, name=None, log_interval=15, default_expiration=None):
+        self._lock = self.__class__.LockType()
+        self._name = name or '{}-{:X}'.format(self.LockType.__name__, id(self))
+        self._default_expiration = default_expiration
+        self._expiration_timer = None
+        self._log_interval = log_interval
+
+        # we want to support both the gevent and builtin lock
+        try:
+            self._lock._owner
```
Comment (idanarye) [inline easypy/concurrency.py]: And PEP8 is OK with this?

File: easypy/concurrency.py (hunk 2 above)
Comment (idanarye) [inline easypy/concurrency.py]: `timeout=-1`? Why not `None`?

File: easypy/concurrency.py [hunk 4]
```diff
@@ -955,6 +958,77 @@ def _get_my_ident():
...
+            return self._lock.acquire(blocking=False)
+
+        log_timer = Timer(expiration=None if timeout < 0 else timeout)
+        while not log_timer.expired:
+            timeout = min(log_timer.remain or self._log_interval, self._log_interval)
+            if self._expiration_timer:
+                timeout = min(self._expiration_timer.remain, timeout)
+            if self._lock.acquire(blocking=True, timeout=timeout):
+                # set the timer after acquiring the lock!
+                self._expiration_timer = self._expiration_timer or Timer(expiration=expiration or self._default_expiration)
```
Comment (idanarye) [inline easypy/concurrency.py]: If this fancy for _create-expiration-timer-if-not-already-exists_?

File: easypy/concurrency.py (hunk 2 above)
Comment (koreno) [inline easypy/concurrency.py]: consistency with python's API

**Pull request merged**